### PR TITLE
WINC-755: Use patched Windows Server 2022 golden image in vSphere CI

### PR DESCRIPTION
**Pull request by saifshaikh48**
Manual backport of relevant commits from #1052:

- Use patched Windows Server 2022 image (with Docker) in vSphere CI
- Update golden image creation docs + scripts with steps to install OS patch
- Announce Windows Server 2022 Support for vSphere

### CODE REVIEW

File: docs/wmco-prerequisites.md [hunk 1]
```diff
@@ -18,11 +18,14 @@ applicable cloud provider.
...
-|----------------|------------------------------------------------------------------------|
-| AWS            | Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2019 |
-| Azure          | Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2019 |
-| VMware vSphere | Windows Server Semi-Annual Channel (SAC): Windows Server 20H2          |
+| Cloud Provider      | Supported Windows Server version                                                     |
+| -----------         | -----------                                                                          |
+| AWS                 | Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2019, version 1809 |
+| Azure               | Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2019, version 1809 |
+| VMware vSphere      | Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2022               |
+|                     | Windows Server Semi-Annual Channel (SAC): Windows Server 20H2                        |
```
Comment (mtnbikenc) [inline docs/wmco-prerequisites.md:26]: When rendered, this line appears in a new row which could be confusing.

File: docs/wmco-prerequisites.md [hunk 2]
```diff
@@ -39,10 +42,11 @@ Note:
...
 
-| Hybrid OVNKubernetes | Supported Windows Server version                                       |
-|----------------------|------------------------------------------------------------------------|
-| Default VXLAN port   | Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2019 |
-| Custom VXLAN port    | Windows Server Semi-Annual Channel (SAC): Windows Server 20H2          |
+| Hybrid OVNKubernetes      | Supported Windows Server version                                                     |
+| -----------               | -----------                                                                          |
+| Default VXLAN port        | Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2019, version 1809 |
+| Custom VXLAN port         | Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2022               |
+|                           | Windows Server Semi-Annual Channel (SAC): Windows Server 20H2                        |
```
Comment (mtnbikenc) [inline docs/wmco-prerequisites.md:49]: Again, when rendered, this line appears in a new row which could be confusing.

File: docs/vsphere_ci/README.md [hunk 3]
```diff
@@ -43,13 +43,14 @@ associated with this public key is what will be used by WMCO to configure VMs cr
 deploying WMCO, this private key will be provided by the user in the form of a Secret.
 
 The [autounattend.xml](scripts/autounattend.xml) file must be edited to update the value of 
-`WindowsPassword` with a user provided password. Then, it executes the following steps:
+`WindowsPassword` with a user provided password. The `ProductKey` must also be updated with a proper value.
+Then, it executes the following steps:
```
Comment (sebsoto) [inline docs/vsphere_ci/README.md:47]: Since this line is being changed anyway, please reword this to be more correct:

autounattend.xml specifies that the following steps should occur:

File: docs/vsphere_ci/README.md [hunk 4]
```diff
@@ -99,6 +100,16 @@ To enable detailed logging:
   PACKER_LOG=1 packer build build.json
 ```
 
+### What to do during the Packer build
+
+During the golden image creation, it is highly recommended to establish access to the virtual machine by launching a
+Web Console through the vSphere Client. This can be done after the Packer build has powered on the VM (while it is
```
Comment (sebsoto) [inline docs/vsphere_ci/README.md:106]: I think this might be more correct?
vSphere Client -> vCenter console

File: docs/vsphere_ci/scripts/install-docker.ps1 [hunk 5]
```diff
@@ -9,3 +9,5 @@ Set-PSRepository PSGallery -InstallationPolicy Trusted
 Install-Module -Name DockerMsftProvider -Repository PSGallery -Force
 # install docker package
 Install-Package -Name docker -ProviderName DockerMsftProvider -Force
```
Comment (sebsoto) [inline docs/vsphere_ci/scripts/install-docker.ps1:11]: Shouldnt we not be doing this for 2022?

**Review by sebsoto — CHANGES_REQUESTED** [review]

**Review by alinaryan — CHANGES_REQUESTED** [review]
HTML works with Markdown so you could list out the server versions by surrounding each OS version with `<ul>OS_Version</ul>`

Example:

| Cloud Provider      | Supported Windows Server version                                                              |
| -----------         | -----------                                                                                   |
| AWS                 | <ul>Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2019, version 1809</ul> |
| Azure               | <ul>Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2019, version 1809</ul> |
| VMware vSphere      | <ul>Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2022</ul><ul>Windows Server Semi-Annual Channel (SAC): Windows Server 20H2</li></ul> |

File: docs/vsphere-golden-image.md [hunk 6]
```diff
@@ -4,13 +4,15 @@ This guide describes the thought process of creating a Windows virtual machine b
...
-required by the operating system to allow using the [hybrid OVN Kubernetes networking with a 
-custom VXLAN port](setup-hybrid-OVNKubernetes-cluster.md#vSphere) feature.
-
-*Please note that Windows Server Long-Term Servicing Channel (LTSC): Windows Server 1809 cannot be used, since 
-the patch is not available.*
+Currently, the Windows Machine Config Operator (WMCO) stable version supports:
+* Windows Server 2022 (must contain the OS-level container networking patch [KB5012637](https://support.microsoft.com/en-us/topic/april-25-2022-kb5012637-os-build-20348-681-preview-2233d69c-d4a5-4be9-8c24-04a450861a8d))
+* Windows Server 20H2
+
+*Please note that Windows Server 2004 and Windows Server 2019, version 1909 (SAC) are no longer supported.
```
Comment (selansen) [inline docs/vsphere-golden-image.md:11]: I prefer to have only the supported version list which we have already.  We don't want to keep adding not supported versions down the line. It's extra work and redundant information.

File: docs/vsphere_ci/README.md [hunk 7]
```diff
@@ -43,13 +43,14 @@ associated with this public key is what will be used by WMCO to configure VMs cr
...
-`WindowsPassword` with a user provided password. Then, it executes the following steps:
+`WindowsPassword` with a user provided password. The `ProductKey` must also be updated with a proper value.
+Then, it executes the following steps:
 
 - Runs `install-vm-tools.cmd` script which installs VMWare tools
 - Runs `configure-vm-tools.ps1` script which configures VMWare tools
 - Runs `install-openssh.ps1` script which installs and configures OpenSSH Server
-- Runs `install-docker.ps1` script which installs Docker
 - Runs `install-firewall-rules.ps1` script which configures the firewall rules
+- Runs `install-docker.ps1` script which installs Docker and restarts the machine
```
Comment (selansen) [inline docs/vsphere_ci/README.md:53]: Now that we merged containerd work already, why do we need to docker?
When we backport this PR in 5.x, we can add this line. WDYT?

File: docs/vsphere_ci/scripts/install-docker.ps1 (hunk 5 above)
Comment (selansen) [reply docs/vsphere_ci/scripts/install-docker.ps1:11]: +1, we should not be doing this.
if we are backporting this into 5.x, then we need to add it backport it in 5.x and other branches that we are trying to backport.

File: docs/vsphere_ci/scripts/install-docker.ps1 [hunk 8]
```diff
@@ -9,3 +9,5 @@ Set-PSRepository PSGallery -InstallationPolicy Trusted
...
+# restart computer after installation as needed for Windows Server 2022
```
[the remark this answers is not in the document]
Comment (selansen) [reply docs/vsphere_ci/scripts/install-docker.ps1:12]: we also need to think about enabling required windows features for container as this is not enabled by default in 2022 server. ( This came out during our Microsoft engineering meeting).
https://github.com/kubernetes-sigs/sig-windows-tools/blob/4598963adbd488e0b7e0eff9f7904c186a122d6c/kubeadm/scripts/Install-Containerd.ps1#L85

File: docs/wmco-prerequisites.md (hunk 1 above)
Comment (selansen) [reply docs/wmco-prerequisites.md:26]: when do we remove 20H2 from the list? After validating  2022 in AWS and AZURE ?

**Review by selansen — CHANGES_REQUESTED** [review]

File: docs/wmco-prerequisites.md (hunk 1 above)
Comment (saifshaikh48) [reply docs/wmco-prerequisites.md:26]: When it goes out of support from MSFT’s end, 2022-08-09

File: hack/machineset.sh [hunk 9]
```diff
@@ -185,7 +185,7 @@ get_vsphere_ms() {
```
Comment (jrvaldes) [inline hack/machineset.sh:185]: You may need to re-open https://github.com/openshift/release/pull/26435

File: docs/vsphere-golden-image.md (hunk 6 above)
Comment (jrvaldes) [reply docs/vsphere-golden-image.md:11]: +1

File: docs/vsphere_ci/scripts/install-docker.ps1 (hunk 8 above)
[the remark this answers is not in the document]
Comment (jrvaldes) [reply docs/vsphere_ci/scripts/install-docker.ps1:12]: Windows features for container are enabled after installing docker (`Install-Package -Name docker -ProviderName DockerMsftProvider -Force`). If we are thinking of removing this step, we may need to add that explicitly.

File: docs/vsphere-golden-image.md [hunk 10]
```diff
@@ -4,13 +4,15 @@ This guide describes the thought process of creating a Windows virtual machine b
...
 
-Currently, the Windows Machine Config Operator (WMCO) stable version only supports Windows Server Semi-Annual
-Channel (SAC): Windows Server 2004, that includes the patch [KB4565351](https://support.microsoft.com/en-us/help/4565351/windows-10-update-kb4565351), 
-required by the operating system to allow using the [hybrid OVN Kubernetes networking with a 
-custom VXLAN port](setup-hybrid-OVNKubernetes-cluster.md#vSphere) feature.
-
-*Please note that Windows Server Long-Term Servicing Channel (LTSC): Windows Server 1809 cannot be used, since 
-the patch is not available.*
+Currently, the Windows Machine Config Operator (WMCO) stable version supports:
+* Windows Server 2022 (must contain the OS-level container networking patch [KB5012637](https://support.microsoft.com/en-us/topic/april-25-2022-kb5012637-os-build-20348-681-preview-2233d69c-d4a5-4be9-8c24-04a450861a8d))
```
Comment (jrvaldes) [inline docs/vsphere-golden-image.md:8]: s/Windows Server 2022/Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2022

File: docs/vsphere_ci/build.json [hunk 11]
```diff
@@ -1,9 +1,9 @@
 {
   "variables": {
-    "os-iso-path": "[WorkloadDatastore] windows-iso-images/winsrv2004_sep_2020_x64_dvd.iso",
+    "os-iso-path": "[WorkloadDatastore] windows-iso-images/winsrv2022_sep_2021_x64_dvd.iso",
```
Comment (jrvaldes) [inline docs/vsphere_ci/build.json:3]: @saifshaikh48   Does this ISO includes the KB5012637 patch?

**Review by jrvaldes — COMMENTED** [review]
@saifshaikh48, 
Thanks for working on this, comments added. PTAL

File: docs/vsphere-golden-image.md (hunk 10 above)
Comment (saifshaikh48) [reply docs/vsphere-golden-image.md:8]: `Windows Server Long-Term Servicing Channel (LTSC): Windows Server 2022` is unnecessarily verbose. I'll use `Windows Server 2022 Long-Term Servicing Channel`

File: docs/vsphere_ci/README.md (hunk 7 above)
Comment (saifshaikh48) [reply docs/vsphere_ci/README.md:53]: I don't think we will backport; we're treating 2022 support as a feature

File: docs/vsphere_ci/README.md (hunk 4 above)
Comment (saifshaikh48) [reply docs/vsphere_ci/README.md:106]: I'm referring to the web page itself, which is a component of a vCenter server

> With vCenter Server, you get access to the new HTML 5 Client

Will change to `vCenter web client` for clarity

File: docs/vsphere_ci/scripts/install-docker.ps1 (hunk 5 above)
Comment (saifshaikh48) [reply docs/vsphere_ci/scripts/install-docker.ps1:11]: Since this is going into 5.1.0, we will have to have docker installed. When docker support is dropped in 6.0.0 then we will have to build a new golden image dropping the docker installation (with [containers feature explicit enabled](https://github.com/openshift/windows-machine-config-operator/pull/1035/commits/1ed065116cf1e0a0eabb7735011a68f3e6f57631?short_path=940babd#r862193771)).

File: hack/machineset.sh (hunk 9 above)
Comment (saifshaikh48) [reply hack/machineset.sh:185]: For sure

File: docs/vsphere_ci/README.md (hunk 7 above)
Comment (saifshaikh48) [reply docs/vsphere_ci/README.md:53]: scratch that -- a 5.x means a release-4.10 backport will be needed

File: docs/vsphere_ci/build.json (hunk 11 above)
Comment (jrvaldes) [reply docs/vsphere_ci/build.json:3]: Then, consider documenting the steps you followed to install the patch. Ideally, it should result in a script executed by Packer during the creation of the VM template.

Comment (saifshaikh48) [reply docs/vsphere_ci/build.json:3]: Sure I can document the steps but automating is not trivial. It's not a default-ly accessible patch (there's some name for it, I forget what they called it in sig-windows meeting) so we have to actually download the files from MSFT and then apply it.

**Comment by saifshaikh48** [comment]
/hold

readme merge conflict

File: docs/vsphere_ci/build.json (hunk 11 above)
Comment (saifshaikh48) [reply docs/vsphere_ci/build.json:3]: [Added](https://github.com/saifshaikh48/windows-machine-config-operator/blob/hotfix-golden-image-2022/docs/vsphere-golden-image.md#5-install-os-level-container-networking-patch-kb5012637)

Comment (jrvaldes) [reply docs/vsphere_ci/build.json:3]: Nice! Wondering if a PowerShell script (e.g. install-kb5012637.ps1) under the [scripts directory](https://github.com/saifshaikh48/windows-machine-config-operator/tree/hotfix-golden-image-2022/docs/vsphere_ci/scripts) that downloads the [file](https://catalog.s.download.windowsupdate.com/c/msdownload/update/software/updt/2022/04/windows10.0-kb5012637-x64_6a7459b60e226b0ad0d30b34a4be069bee4d2867.msu) can be included in the [sysprep](https://github.com/saifshaikh48/windows-machine-config-operator/blob/hotfix-golden-image-2022/docs/vsphere_ci/scripts/autounattend.xml#L101) process. Will that capture the automation of the above [steps](https://github.com/saifshaikh48/windows-machine-config-operator/blob/hotfix-golden-image-2022/docs/vsphere-golden-image.md#5-install-os-level-container-networking-patch-kb5012637).

**Comment by saifshaikh48** [comment]
/hold

https://github.com/openshift/windows-machine-config-operator/pull/1052 to go in first

**Comment by saifshaikh48** [comment]
/test vsphere-e2e-operator

changed name of template for use in release-4.10 to `windows-server-2022-template-withDocker` as no pre-installed Docker is the default in master and beyond

**Comment by saifshaikh48** [comment]
/hold

vsphere job landed in IBM cloud environment. Need to engage splat to get the golden image replicated there

**Comment by saifshaikh48** [comment]
/unhold

https://github.com/openshift/windows-machine-config-operator/pull/1052 has merged
